door-runtime/src/jsMain/kotlin/wrappers/SQLiteDatasourceJs.kt

```diff
@@ -39,29 +56,70 @@ class SQLiteDatasourceJs(private val dbName: String, private val worker: Worker)
         val completable = CompletableDeferred<WorkerResult>()
         val actionId = ++idCounter
         pendingMessages[actionId] = completable
+        executedSqlQueries[actionId] = message["sql"].toString()
         message["id"] = actionId
         worker.postMessage(message)
         return completable.await()
     }
 
-    private fun makeMessage(sql: String, params: Array<Any?>?) : Json {
-        val message = json(
+    private fun makeMessage(sql: String, params: Array<Any?>? = arrayOf()): Json {
+        return json(
             "action" to "exec",
-            "sql" to sql
+            "sql" to sql,
+            "params" to JSON.parse(
+                JSON.stringify(params?.map { (if(it != js("undefined")) it else null).toString()
+            }))
         )
-
-        if(params != null) {
-            message["params"] = params
-        }
-        return message
     }
 
-    internal suspend fun sendQuery(sql: String, params: Array<Any?>?): ResultSet {
+    internal suspend fun sendQuery(sql: String, params: Array<Any?>? = null): ResultSet {
         return sendMessage(makeMessage(sql, params)).results?.let { SQLiteResultSet(it) } ?: SQLiteResultSet(arrayOf())
     }
 
-    internal suspend fun sendUpdate(sql: String, params: Array<Any?>?): Int {
-        return sendMessage(makeMessage(sql, params)).let { if(it.ready) 1 else 0 }
+    internal suspend fun sendUpdate(sql: String, params: Array<Any?>?, returnGeneratedKey: Boolean = false): Int {
+        val newSql = makeSqlQuery(sql,params, returnGeneratedKey)
+        val workerResult = sendMessage(makeMessage(newSql, if(returnGeneratedKey) arrayOf() else params))
+        val results = workerResult.results
+        if(results != null){
+            generatedKeys = SQLiteResultSet(results)
+        }
+        return workerResult.let { if(it.ready) 1 else 0 }
+    }
+
+
+    /**
+     * SQL JS doesn't work when you try to execute multiple queries at a time with one of them having params.
+     * It will try to bind params event to that query with no params which will result to column
+     * index out of range exception.
+     *
+     * Instead, we generate new SQL query with inline values so that we can execute those queries without passing
+     * any params.
+     *
+     * NOTE:
+     * SQL inline values is known for slowing down SQL performance.
+     */
+    private fun makeSqlQuery(sql: String, params: Array<Any?>?, returnGeneratedKey: Boolean): String {
```

Comment:
Change to using a lock to wait for the reply. If a generated key result is expected, then the query to get the generated key before unlocking.

Reply:
e.g.

/**
     * Execute SQL task by sending a message via Worker
     * @param message message to be sent for SQLJs to execute
     */
    suspend fun sendMessage(message: Json): WorkerResult {
        lock.withLock {
            val completable = CompletableDeferred<WorkerResult>()
            val actionId = ++idCounter
            pendingMessages[actionId] = completable
            executedSqlQueries[actionId] = message["sql"].toString()
            message["id"] = actionId
            worker.postMessage(message)
            val result = completable.await()
            
            //if needed... get the generated key
            return result
        }
        
    }